### PR TITLE
Infer QueryBuilderType for any method returning QB

### PR DESCRIPTION
**Pull request by janedbal**
Resolves https://github.com/phpstan/phpstan-doctrine/issues/499

### CODE REVIEW

File: src/Type/Doctrine/QueryBuilder/QueryBuilderReturnedDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	public function __construct(
+		OtherMethodQueryBuilderParser $otherMethodQueryBuilderParser
+	)
+	{
+		$this->otherMethodQueryBuilderParser = $otherMethodQueryBuilderParser;
+	}
+
+	public function getClass(): string
+	{
+		return 'QueryResult\CreateQuery\QueryBuilderGetQuery'; // TODO https://github.com/phpstan/phpstan-src/pull/2761
```
Comment (janedbal) [inline src/Type/Doctrine/QueryBuilder/QueryBuilderReturnedDynamicReturnTypeExtension.php]: Change to 'object' once merged.

Comment (janedbal) [reply src/Type/Doctrine/QueryBuilder/QueryBuilderReturnedDynamicReturnTypeExtension.php]: Outbuilt by ExpressionTypeResolverExtension

File: src/Type/Doctrine/QueryBuilder/ReturnQueryBuilderExpressionTypeResolverExtension.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		}
+
+		$callerType = $scope->getType($expr->var);
+
+		foreach ($callerType->getObjectClassReflections() as $callerClassReflection) {
+			if ($callerClassReflection->is(QueryBuilder::class)) {
+				return null; // covered by QueryBuilderMethodDynamicReturnTypeExtension
+			}
+			if ($callerClassReflection->is(EntityRepository::class)) {
+				return null; // createQueryBuilder covered by EntityRepositoryCreateQueryBuilderDynamicReturnTypeExtension
```
Comment (janedbal) [inline src/Type/Doctrine/QueryBuilder/ReturnQueryBuilderExpressionTypeResolverExtension.php]: Ignore only such method, people can add custom ones.

File: src/Type/Doctrine/QueryBuilder/ReturnQueryBuilderExpressionTypeResolverExtension.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	public function __construct(
+		OtherMethodQueryBuilderParser $otherMethodQueryBuilderParser
+	)
+	{
+		$this->otherMethodQueryBuilderParser = $otherMethodQueryBuilderParser;
+	}
+
+	public function getType(Expr $expr, Scope $scope): ?Type
+	{
+		if (!$expr instanceof MethodCall) {
```
Comment (janedbal) [inline src/Type/Doctrine/QueryBuilder/ReturnQueryBuilderExpressionTypeResolverExtension.php]: Support static calls.

**Comment by janedbal** [comment]
@ondrejmirtes I think this is reviewable now. I tested this on our big codebase and it seems to be working great.

**Comment by ondrejmirtes** [comment]
Thank you. I'm gonna test it and release it.

**Pull request merged**